eval: refactor builtins that access Context.Txn explicitly

yuzefovich opened the pull request:
This commit extracts all builtins that access `eval.Context.Txn` field
explicitly into a newly-added `FnWithTxn` field of the overload. This is
a step towards removing `eval.Context.Txn` field to clarify which txn is
being used during the execution.

`Context` is type-aliased to `evaluator` which implements a few
evaluator interfaces. This commit introduces an extension that overrides
`EvalFuncExpr` method implementation to handle `FnWithTxn` overloads,
everything else is delegated to `evaluator` (the "vanilla" evaluator
returns an error when trying to evaluate `FnWithTxn`, which is an
assertion failure in test builds).

`eval.Expr` signature has been adjusted to take in an optional txn
argument. If expression evaluation is performed in the context where we
_might_ be evaluating builtins that use txns, then the txn needs to be
given or explicit `nil` txn needs to be passed to use the extended
`evaluatorWithTxn`.

Informs: #41992.
Epic: None
Release note: None

yuzefovich committed:
WIP on removing eval.Context.Txn

yuzefovich committed:
eval: refactor builtins that access Context.Txn explicitly

This commit extracts all builtins that access `eval.Context.Txn` field
explicitly into a newly-added `FnWithTxn` field of the overload. This is
a step towards removing `eval.Context.Txn` field to clarify which txn is
being used during the execution.

`Context` is type-aliased to `evaluator` which implements a few
evaluator interfaces. This commit introduces an extension that overrides
`EvalFuncExpr` method implementation to handle `FnWithTxn` overloads,
everything else is delegated to `evaluator` (the "vanilla" evaluator
returns an error when trying to evaluate `FnWithTxn`, which is an
assertion failure in test builds).

`eval.Expr` signature has been adjusted to take in an optional txn
argument. If expression evaluation is performed in the context where we
_might_ be evaluating builtins that use txns, then the txn needs to be
given or explicit `nil` txn needs to be passed to use the extended
`evaluatorWithTxn`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147738)
<!-- Reviewable:end -->

yuzefovich committed:
WIP on making the previous commit work

yuzefovich committed:
WIP on plumbing txn into optbuilder.Builder

WIP on plumbing txn into norm.Factory

The main idea is to follow where `evalCtx` is instantiated.

yuzefovich committed:
WIP on cleaning up FlowCtx a bit
